### PR TITLE
Standardize scCODA plot palette interface

### PR DESCRIPTION
**Pull request by mschilli87**
**PR Checklist**

-   [x] Referenced issue is linked
-   [ ] If you've fixed a bug or added code that should be tested, add tests!
-   [ ] Documentation in `docs` is updated

**Description of changes**

This commit ensures that every scCODA plot function with a `palette` parameter accepts at least a string value (color map name) and a `matplotlib.colors.ListedColormap` object (color map). The default behaviour remains unchanged and the new function interfaces are backwards-compatible.

**Technical details**

These changed code is untested.

I would appreciate some guidance regarding how to add tests covering the new code and update the documentaion.

In the meantime, I leave this as a draft.

**Additional context**

This fixes https://github.com/scverse/pertpy/issues/771.

### CODE REVIEW

**Comment by mschilli87** [comment]
@Zethson: To me the failing tests look unrelated to my changes. If you could have a look to double-check and advice me how to move this forward, I'd appreciate your input.

**Comment by mschilli87** [comment]
Note: The current version does not appear to work:

```python
import pertpy as pt
from matplotlib import cm

haber_cells = pt.dt.haber_2017_regions()

sccoda_model = pt.tl.Sccoda()
sccoda_data = sccoda_model.load(
    haber_cells,
    type="cell_level",
    generate_sample_level=True,
    cell_type_identifier="cell_label",
    sample_identifier="batch",
    covariate_obs=["condition"],
)

sccoda_data.mod["coda_salm"] = sccoda_data["coda"][
    sccoda_data["coda"].obs["condition"].isin(["Control", "Salmonella"])
].copy()

sccoda_model.plot_boxplots(sccoda_data, modality_key="coda_salm",
                           feature_name="condition", add_dots=True)
sccoda_model.plot_boxplots(sccoda_data, modality_key="coda_salm",
                           feature_name="condition", add_dots=True,
                           palette="BuPu")
sccoda_model.plot_boxplots(sccoda_data, modality_key="coda_salm",
                           feature_name="condition", add_dots=True,
                           palette=cm.BuPu)
```

```pytb
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<path>/pertpy/tools/_coda/_base_coda.py", line 1646, in plot_boxplots
    ax = sns.boxplot(
         ^^^^^^^^^^^^
  File "<path>/.pertpy.venv/lib/python3.12/site-packages/seaborn/categorical.py", line 1626, in boxplot
    p.map_hue(palette=palette, order=hue_order, norm=hue_norm, saturation=saturation)
  File "<path>/.pertpy.venv/lib/python3.12/site-packages/seaborn/_base.py", line 838, in map_hue
    mapping = HueMapping(self, palette, order, norm, saturation)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/.pertpy.venv/lib/python3.12/site-packages/seaborn/_base.py", line 150, in __init__
    levels, lookup_table = self.categorical_mapping(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/.pertpy.venv/lib/python3.12/site-packages/seaborn/_base.py", line 248, in categorical_mapping
    colors = color_palette(palette, n_colors)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/.pertpy.venv/lib/python3.12/site-packages/seaborn/palettes.py", line 245, in color_palette
    pal_cycle = cycle(palette)
                ^^^^^^^^^^^^^^
TypeError: 'LinearSegmentedColormap' object is not iterable
```

I'll work on getting this to work locally (and similarly for the other plot functions) and will report back.

**Review by Zethson — COMMENTED** [review]
Thank you!

Yes, only the pre-release CI is failing. I opened an issue for this https://github.com/scverse/pertpy/issues/774. It's unrelated to your PR.

As long as the normal 3.13 CI passes, we're good.

Testing plots is tricky business. This is another shortcut that we took with pertpy. This is what plot tests could look like: https://github.com/theislab/ehrapy/tree/main/tests/plot . However, since we haven't done it systematically yet, I do not expect it from you for this PR. Simply demonstrating it with a few small reproducible examples will be enough.

**Comment by mschilli87** [comment]
Perfect! ❤️ 
I already fixed the above issue locally and will now try similar tests for the other functions before (force) pushing and sharing my test code for you to reproduce.

**Comment by Zethson** [comment]
Thank you very much!

We have an example for `plot_effects_umap` here: https://pertpy.readthedocs.io/en/stable/api/tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda.plot_effects_umap . I hope that it works at least. In any case, I understand that you either found a bug or terrible documentation/UX/error messages as you got confused. Another one for the backlog.

File: pertpy/tools/_coda/_base_coda.py [hunk 1]
```diff
@@ -1535,6 +1537,8 @@ def plot_boxplots(  # pragma: no cover # noqa: D417
             args_swarmplot = {}
         if isinstance(data, MuData):
             data = data[modality_key]
+        if isinstance(palette, Colormap):
+            palette = np.array([palette(i % palette.N) for i in range(2)]).tolist()
```
Comment (Zethson) [inline pertpy/tools/_coda/_base_coda.py:1541]: Not the most easy version and I wonder whether just

`palette = [palette(i) for i in range(2)]`

would be enough?

File: pertpy/tools/_coda/_base_coda.py [hunk 2]
```diff
@@ -2173,6 +2177,11 @@ def plot_effects_umap(  # pragma: no cover # noqa: D417
         data_coda = mdata[modality_key_2]
         if isinstance(effect_name, str):
             effect_name = [effect_name]
+        if isinstance(palette, Colormap):
+            palette = {
+                cluster: palette(i % palette.N)
+                for i, cluster in enumerate(data_rna.obs[cluster_key].unqiue().to_list())
```
Comment (Zethson) [inline pertpy/tools/_coda/_base_coda.py:2183]:
```suggestion
                for i, cluster in enumerate(data_rna.obs[cluster_key].unique().to_list())
```

given that this typo was in the code, do your example cases actually cover this?

**Review by Zethson — CHANGES_REQUESTED** [review]
Almost ready to get merged. Thank you!

File: pertpy/tools/_coda/_base_coda.py (hunk 2 above)
Comment (mschilli87) [reply pertpy/tools/_coda/_base_coda.py:2183]: Good question. I will double-check.😅

Comment (mschilli87) [reply pertpy/tools/_coda/_base_coda.py:2183]: It was not as it is in `plot_effects_umap` which I failed to get working without setting a `palette`.

File: pertpy/tools/_coda/_base_coda.py (hunk 1 above)
Comment (mschilli87) [reply pertpy/tools/_coda/_base_coda.py:1541]: I actually think if we assume `palette.N > 1`, `palette = palette(range(2))` should do.

File: pertpy/tools/_coda/_base_coda.py (hunk 2 above)
Comment (Zethson) [reply pertpy/tools/_coda/_base_coda.py:2183]: Right, I see. Then this is on my plate :)

File: pertpy/tools/_coda/_base_coda.py (hunk 1 above)
Comment (Zethson) [reply pertpy/tools/_coda/_base_coda.py:1541]: Okie! If you test that quickly and could push it, I'm happy to merge the PR.

**Comment by mschilli87** [comment]
@Zethson
> We have an example for `plot_effects_umap` here: https://pertpy.readthedocs.io/en/stable/api/tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda.plot_effects_umap.

I saw that but didn't feel like installing running tascCODA just to test this.
I was hoping I missed something but I guess now I have to. 😉

**Comment by Zethson** [comment]
You might also require schist which is only available on conda. If you can't get it to work, I'll have a look in the next 2 days.

It's both super helpful and embarassing that you find all of these issues 😄

**Comment by mschilli87** [comment]
Pertpy contains so many things with even more parameters. I think it is normal to miss edge cases as long as you don't have either a lot of end users and/or a very solid CI/test setup.

I am happy to help. I get a lot from FOSS research tools. I am happy if I can contribute back rather than hacking around issues in my own playground.

**Comment by mschilli87** [comment]
@Zethson
> You might also require schist which is only available on conda.

Thank you for the advice.
I setup a conda env with schist and build pertpy locally within in so I coud test the example code you linked:

```py
import pertpy as pt
import scanpy as sc
import schist
adata = pt.dt.haber_2017_regions()
sc.pp.neighbors(adata)
schist.inference.nested_model(adata, n_init=100, random_seed=5678)
tasccoda_model = pt.tl.Tasccoda()
tasccoda_data = tasccoda_model.load(adata, type="cell_level",
                cell_type_identifier="nsbm_level_1",
                sample_identifier="batch", covariate_obs=["condition"],
                levels_orig=["nsbm_level_4", "nsbm_level_3", "nsbm_level_2", "nsbm_level_1"],
                add_level_name=True)
tasccoda_model.prepare(
    tasccoda_data,
    modality_key="coda",
    reference_cell_type="18",
    formula="condition",
    pen_args=dict(phi=0, lambda_1=3.5),
    tree_key="tree"
)
tasccoda_model.run_nuts(
    tasccoda_data, modality_key="coda", rng_key=1234, num_samples=10000, num_warmup=1000
)
tasccoda_model.run_nuts(
    tasccoda_data, modality_key="coda", rng_key=1234, num_samples=10000, num_warmup=1000
)
sc.tl.umap(tasccoda_data["rna"])
tasccoda_model.plot_effects_umap(tasccoda_data,
                        effect_name=["effect_df_condition[T.Salmonella]",
                                     "effect_df_condition[T.Hpoly.Day3]",
                                     "effect_df_condition[T.Hpoly.Day10]"],
                                      cluster_key="nsbm_level_1",
                        )
```

I don't understand why it contains the

```python
tasccoda_model.run_nuts(
    tasccoda_data, modality_key="coda", rng_key=1234, num_samples=10000, num_warmup=1000
)
```

line twice (especially given that this step takes about 10 min), but at this point I just ran it as-is and and least this way I got to run [`pt.tl.Sccoda.plot_effects_umap`](https://pertpy.readthedocs.io/en/stable/api/tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda.plot_effects_umap) without an error.

Thus, I tested my changes by running

```python
from matplotlib import cm

tasccoda_model.plot_effects_umap(tasccoda_data,
                                 effect_name=["effect_df_condition[T.Salmonella]",
                                              "effect_df_condition[T.Hpoly.Day3]",
                                              "effect_df_condition[T.Hpoly.Day10]"],
                                 cluster_key="nsbm_level_1",
                                 palette="BuPu")
tasccoda_model.plot_effects_umap(tasccoda_data,
                                 effect_name=["effect_df_condition[T.Salmonella]",
                                              "effect_df_condition[T.Hpoly.Day3]",
                                              "effect_df_condition[T.Hpoly.Day10]"],
                                 cluster_key="nsbm_level_1",
                                 palette=cm.BuPu)
```

Which didn't error but gave me a `FutureWarning`:

```
<path>/pertpy/tools/_coda/_base_coda.py:2183: FutureWarning: Categorical.to_list is deprecated and will be removed in a future version. Use obj.tolist() instead
  for i, cluster in enumerate(data_rna.obs[cluster_key].unique().to_list())
```

At least we know for sure my test covered this line now. 😅
I fixed this but will run another test round just to be sure before pushing.

**Comment by mschilli87** [comment]
For my re-run I also tested running `tasccoda_model.run_nuts` only once and it turned out to be sufficient. 😉

This is the final full test I ran (in addition to the one [above](https://github.com/scverse/pertpy/pull/773#issuecomment-2897004152) that still runs with the same result):

```python
import pertpy as pt
import scanpy as sc
import schist
from matplotlib import cm

adata = pt.dt.haber_2017_regions()

sc.pp.neighbors(adata)
schist.inference.nested_model(adata, n_init=100, random_seed=5678)

tasccoda_model = pt.tl.Tasccoda()
tasccoda_data = tasccoda_model.load(
    adata, type="cell_level", cell_type_identifier="nsbm_level_1",
    sample_identifier="batch", covariate_obs=["condition"],
    levels_orig=["nsbm_level_4", "nsbm_level_3", "nsbm_level_2", "nsbm_level_1"],
    add_level_name=True
)

tasccoda_model.prepare(tasccoda_data, modality_key="coda",
                       reference_cell_type="18",
                       formula="condition",
                       pen_args=dict(phi=0, lambda_1=3.5),
                       tree_key="tree")

tasccoda_model.run_nuts(tasccoda_data, modality_key="coda", rng_key=1234,
                        num_samples=10000, num_warmup=1000)

sc.tl.umap(tasccoda_data["rna"])

tasccoda_model.plot_effects_umap(tasccoda_data,
                                 effect_name=["effect_df_condition[T.Salmonella]",
                                              "effect_df_condition[T.Hpoly.Day3]",
                                              "effect_df_condition[T.Hpoly.Day10]"],
                                 cluster_key="nsbm_level_1")
tasccoda_model.plot_effects_umap(tasccoda_data,
                                 effect_name=["effect_df_condition[T.Salmonella]",
                                              "effect_df_condition[T.Hpoly.Day3]",
                                              "effect_df_condition[T.Hpoly.Day10]"],
                                 cluster_key="nsbm_level_1",
                                 palette="BuPu")
tasccoda_model.plot_effects_umap(tasccoda_data,
                                 effect_name=["effect_df_condition[T.Salmonella]",
                                              "effect_df_condition[T.Hpoly.Day3]",
                                              "effect_df_condition[T.Hpoly.Day10]"],
                                 cluster_key="nsbm_level_1",
                                 palette=cm.BuPu)
```

With this, I think my PR is good but the docs of [`plot_effects_umap`](https://pertpy.readthedocs.io/en/stable/api/tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda.plot_effects_umap) need some fixing and probably it should be possible to use this function for scCODA effects as well, not just after runnning tascCODA with all it's additional requirements.
Let me know if you want me to open separate issues for those.

**Comment by Zethson** [comment]
> With this, I think my PR is good but the docs of [plot_effects_umap](https://pertpy.readthedocs.io/en/stable/api/tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda.plot_effects_umap) need some fixing and probably it should be possible to use this function for scCODA effects as well, not just after running tascCODA with all it's additional requirements.
Let me know if you want me to open separate issues for those.

Yes, please and if you're willing to help here, I'd of course be very happy.

**Review by Zethson — APPROVED** [review]
Under the assumption that we only need 2 colors for one of the cases, this should be good now.

Thank you!
